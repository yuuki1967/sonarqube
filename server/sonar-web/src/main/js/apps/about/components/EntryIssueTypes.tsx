/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { Link } from 'react-router';
import { formatMeasure } from '../../../helpers/measures';
import { translate } from '../../../helpers/l10n';
import { getIssuesUrl } from '../../../helpers/urls';
import BugIcon from '../../../components/icons-components/BugIcon';
import VulnerabilityIcon from '../../../components/icons-components/VulnerabilityIcon';
import CodeSmellIcon from '../../../components/icons-components/CodeSmellIcon';

interface Props {
  bugs?: number;
  codeSmells?: number;
  loading: boolean;
  vulnerabilities?: number;
}

export default function EntryIssueTypes({ bugs, codeSmells, loading, vulnerabilities }: Props) {
  return (
    <div className="about-page-projects">
      {loading ? (
        <i className="spinner" />
      ) : (
        <table className="about-page-issue-types">
          <tbody>
            <tr>
              <td className="about-page-issue-type-number">
                <Link
                  className="about-page-issue-type-link"
                  to={getIssuesUrl({ resolved: 'false', types: 'BUG', s: 'CREATION_DATE' })}>
                  {formatMeasure(bugs, 'SHORT_INT')}
                </Link>
              </td>
              <td>
                <span className="little-spacer-right">
                  <BugIcon />
                </span>
                {translate('issue.type.BUG.plural')}
              </td>
            </tr>
            <tr>
              <td className="about-page-issue-type-number">
                <Link
                  className="about-page-issue-type-link"
                  to={getIssuesUrl({
                    resolved: 'false',
                    types: 'VULNERABILITY',
                    s: 'CREATION_DATE'
                  })}>
                  {formatMeasure(vulnerabilities, 'SHORT_INT')}
                </Link>
              </td>
              <td>
                <span className="little-spacer-right">
                  <VulnerabilityIcon />
                </span>
                {translate('issue.type.VULNERABILITY.plural')}
              </td>
            </tr>
            <tr>
              <td className="about-page-issue-type-number">
                <Link
                  className="about-page-issue-type-link"
                  to={getIssuesUrl({ resolved: 'false', types: 'CODE_SMELL', s: 'CREATION_DATE' })}>
                  {formatMeasure(codeSmells, 'SHORT_INT')}
                </Link>
              </td>
              <td>
                <span className="little-spacer-right">
                  <CodeSmellIcon />
                </span>
                {translate('issue.type.CODE_SMELL.plural')}
              </td>
            </tr>
          </tbody>
        </table>
      )}
    </div>
  );
}
