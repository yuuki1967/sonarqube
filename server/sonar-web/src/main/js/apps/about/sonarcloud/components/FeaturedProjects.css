/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.sc-featured-projects {
  margin-top: 40px;
  position: relative;
}

.sc-featured-projects-container {
  margin: 0 40px;
  overflow: hidden;
  padding: 10px 8px 20px;
  max-width: 985px;
}

.sc-featured-projects-inner {
  display: flex;
  order: 2;
  position: relative;
  left: calc(100% / -3);
  transform: translateX(calc(100% / 3));
}

.sc-featured-projects-inner.reversing {
  transform: translateX(calc(100% / -3));
}

.sc-featured-projects-inner.ready {
  transform: translateX(0);
  transition: transform 0.5s cubic-bezier(0.23, 1, 0.32, 1);
}

.sc-project-card-container {
  flex: 1 0 calc(100% / 3);
}

.sc-project-card {
  width: 255px;
  display: inline-block;
  margin: 0 10px;
  padding: 25px;
  border: 1px solid var(--sonarcloudBlack300) !important;
  border-radius: 5px;
  box-shadow: 0 1px 1px rgba(7, 7, 6, 0.1);
  transition: all 0.1s ease-in;
  color: inherit;
}

.sc-project-card:hover,
.sc-project-card:focus {
  box-shadow: 0 10px 30px rgba(7, 7, 6, 0.2);
  transform: translateY(-4px);
  color: inherit;
}

.sc-project-card-header {
  padding: 0 10px 16px;
}

.sc-project-card-limited {
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.sc-project-card-measures li {
  display: flex;
  align-items: baseline;
  justify-content: space-between;
  padding: 6px 0 4px;
  border-bottom: 1px solid var(--sonarcloudBlack250);
}

.sc-project-card-measures li:last-of-type {
  border: none;
}

.sc-project-card-measures li div {
  display: inline-flex;
}

.sc-project-card .sc-mention {
  color: var(--sonarcloudBlack500);
}

.sc-project-button {
  display: block;
  position: absolute;
  top: 50%;
  margin: 0;
  padding: 4px;
  border: none;
  background: transparent;
  color: var(--sonarcloudBlack300);
  text-decoration: none;
  cursor: pointer;
  outline: none;
  transition: all 0.2s ease;
}

.sc-project-button img {
  opacity: 0.5;
  width: 19px;
}

.sc-project-button:hover img {
  opacity: 1;
}

.sc-project-button:last-child {
  right: 0;
}

.sc-project-button:hover,
.sc-project-button:active,
.sc-project-button:focus {
  color: var(--sonarcloudBlack500);
}
