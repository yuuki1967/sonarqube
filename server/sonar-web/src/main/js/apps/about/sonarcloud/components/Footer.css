/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.sc-footer {
  background-color: var(--sonarcloudBlack800);
  color: var(--sonarcloudBlack300);
  font-size: 12px;
  min-width: var(--minPageWidth);
}

.sc-footer *:focus {
  box-shadow: 0 0 0 3px rgba(230, 92, 0, 0.25);
}

.sc-footer-limited {
  position: relative;
  max-width: 1280px;
  margin-left: auto;
  margin-right: auto;
  padding-left: 20px;
  padding-right: 20px;
}

.sc-footer-copy {
  padding: 20px 0;
  border-top: 1px solid var(--sonarcloudBlack700);
  line-height: 16px;
  font-size: 11px;
  text-align: center;
}

.sc-footer-copy-link {
  text-decoration: underline;
}

.sc-footer-nav {
  display: flex;
  padding: 30px 0;
}

.sc-footer-nav-column {
  margin-right: 60px;
}

.sc-footer-nav-column-title {
  line-height: 1;
  margin-bottom: 16px;
  color: #fff;
  font-size: 12px;
  font-weight: 700;
  text-transform: uppercase;
}

.sc-footer-link {
  border: none;
  color: inherit;
}

.sc-footer-link:hover {
  color: #fff;
}

.sc-footer-link:focus {
  color: var(--sonarcloudOrange500);
}

.sc-footer-logo {
  position: absolute;
  top: 30px;
  right: 20px;
}
