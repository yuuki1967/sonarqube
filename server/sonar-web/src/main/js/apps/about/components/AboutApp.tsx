/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { connect } from 'react-redux';
import { keyBy } from 'lodash';
import { Link } from 'react-router';
import { Location } from 'history';
import AboutProjects from './AboutProjects';
import AboutLanguages from './AboutLanguages';
import AboutCleanCode from './AboutCleanCode';
import AboutQualityModel from './AboutQualityModel';
import AboutQualityGates from './AboutQualityGates';
import AboutLeakPeriod from './AboutLeakPeriod';
import AboutStandards from './AboutStandards';
import AboutScanners from './AboutScanners';
import EntryIssueTypes from './EntryIssueTypes';
import GlobalContainer from '../../../app/components/GlobalContainer';
import { searchProjects } from '../../../api/components';
import { getFacet } from '../../../api/issues';
import { fetchAboutPageSettings } from '../actions';
import {
  getAppState,
  getCurrentUser,
  getGlobalSettingValue,
  Store
} from '../../../store/rootReducer';
import { translate } from '../../../helpers/l10n';
import { addWhitePageClass, removeWhitePageClass } from '../../../helpers/pages';
import '../styles.css';

interface Props {
  appState: Pick<T.AppState, 'defaultOrganization' | 'organizationsEnabled'>;
  currentUser: T.CurrentUser;
  customText?: string;
  fetchAboutPageSettings: () => Promise<void>;
  location: Location;
}

interface State {
  issueTypes?: { [key: string]: { count: number } };
  loading: boolean;
  projectsCount: number;
}

class AboutApp extends React.PureComponent<Props, State> {
  mounted = false;

  state: State = {
    loading: true,
    projectsCount: 0
  };

  componentDidMount() {
    this.mounted = true;
    this.loadData();
    addWhitePageClass();
  }

  componentWillUnmount() {
    this.mounted = false;
    removeWhitePageClass();
  }

  loadProjects() {
    return searchProjects({ ps: 1 }).then(r => r.paging.total);
  }

  loadIssues() {
    return getFacet({ resolved: false }, 'types');
  }

  loadCustomText() {
    return this.props.fetchAboutPageSettings();
  }

  loadData() {
    Promise.all([this.loadProjects(), this.loadIssues(), this.loadCustomText()]).then(
      responses => {
        if (this.mounted) {
          const [projectsCount, issues] = responses;
          const issueTypes = keyBy(issues.facet, 'val');
          this.setState({ projectsCount, issueTypes, loading: false });
        }
      },
      () => {
        if (this.mounted) {
          this.setState({ loading: false });
        }
      }
    );
  }

  render() {
    const { customText } = this.props;
    const { loading, issueTypes, projectsCount } = this.state;

    let bugs;
    let vulnerabilities;
    let codeSmells;
    if (!loading && issueTypes) {
      bugs = issueTypes['BUG'] && issueTypes['BUG'].count;
      vulnerabilities = issueTypes['VULNERABILITY'] && issueTypes['VULNERABILITY'].count;
      codeSmells = issueTypes['CODE_SMELL'] && issueTypes['CODE_SMELL'].count;
    }

    return (
      <GlobalContainer location={this.props.location}>
        <div className="page page-limited about-page" id="about-page">
          <div className="about-page-entry">
            <div className="about-page-intro">
              <h1 className="big-spacer-bottom">{translate('layout.sonar.slogan')}</h1>
              {!this.props.currentUser.isLoggedIn && (
                <Link className="button button-active big-spacer-right" to="/sessions/new">
                  {translate('layout.login')}
                </Link>
              )}
              <a
                className="button"
                href="https://redirect.sonarsource.com/doc/home.html"
                rel="noopener noreferrer"
                target="_blank">
                {translate('about_page.read_documentation')}
              </a>
            </div>

            <div className="about-page-instance">
              <AboutProjects count={projectsCount} loading={loading} />
              <EntryIssueTypes
                bugs={bugs}
                codeSmells={codeSmells}
                loading={loading}
                vulnerabilities={vulnerabilities}
              />
            </div>
          </div>

          {customText && (
            <div className="about-page-section" dangerouslySetInnerHTML={{ __html: customText }} />
          )}

          <AboutLanguages />

          <AboutQualityModel />

          <div className="flex-columns">
            <div className="flex-column flex-column-half about-page-group-boxes">
              <AboutCleanCode />
            </div>
            <div className="flex-column flex-column-half about-page-group-boxes">
              <AboutLeakPeriod />
            </div>
          </div>

          <div className="flex-columns">
            <div className="flex-column flex-column-half about-page-group-boxes">
              <AboutQualityGates />
            </div>
            <div className="flex-column flex-column-half about-page-group-boxes">
              <AboutStandards appState={this.props.appState} />
            </div>
          </div>

          <AboutScanners />
        </div>
      </GlobalContainer>
    );
  }
}

const mapStateToProps = (state: Store) => {
  const customText = getGlobalSettingValue(state, 'sonar.lf.aboutText');
  return {
    appState: getAppState(state),
    currentUser: getCurrentUser(state),
    customText: customText && customText.value
  };
};

const mapDispatchToProps = { fetchAboutPageSettings } as any;

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AboutApp);
