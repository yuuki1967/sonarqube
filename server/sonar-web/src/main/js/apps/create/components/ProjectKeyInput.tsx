/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as classNames from 'classnames';
import { debounce } from 'lodash';
import ValidationInput from '../../../components/controls/ValidationInput';
import { doesComponentExists } from '../../../api/components';
import { translate } from '../../../helpers/l10n';

interface Props {
  className?: string;
  initialValue?: string;
  onChange: (value: string | undefined) => void;
}

interface State {
  error?: string;
  touched: boolean;
  validating: boolean;
  value: string;
}

export default class ProjectKeyInput extends React.PureComponent<Props, State> {
  mounted = false;
  constructor(props: Props) {
    super(props);
    this.state = { error: undefined, touched: false, validating: false, value: '' };
    this.checkFreeKey = debounce(this.checkFreeKey, 250);
  }

  componentDidMount() {
    this.mounted = true;
    if (this.props.initialValue !== undefined) {
      this.setState({ value: this.props.initialValue });
      this.validateKey(this.props.initialValue);
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  checkFreeKey = (key: string) => {
    this.setState({ validating: true });
    return doesComponentExists({ component: key })
      .then(alreadyExist => {
        if (this.mounted) {
          if (!alreadyExist) {
            this.setState({ error: undefined, validating: false });
            this.props.onChange(key);
          } else {
            this.setState({
              error: translate('onboarding.create_project.project_key.taken'),
              touched: true,
              validating: false
            });
            this.props.onChange(undefined);
          }
        }
      })
      .catch(() => {
        if (this.mounted) {
          this.setState({ error: undefined, validating: false });
          this.props.onChange(key);
        }
      });
  };

  handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = event.currentTarget;
    this.setState({ touched: true, value });
    this.validateKey(value);
  };

  validateKey(key: string) {
    if (key.length > 400 || !/^[\w-.:]*[a-zA-Z]+[\w-.:]*$/.test(key)) {
      this.setState({
        error: translate('onboarding.create_project.project_key.error'),
        touched: true
      });
      this.props.onChange(undefined);
    } else {
      this.checkFreeKey(key);
    }
  }

  render() {
    const isInvalid = this.state.touched && this.state.error !== undefined;
    const isValid = this.state.touched && !this.state.validating && this.state.error === undefined;
    return (
      <ValidationInput
        className={this.props.className}
        description={translate('onboarding.create_project.project_key.description')}
        error={this.state.error}
        help={translate('onboarding.create_project.project_key.help')}
        id="project-key"
        isInvalid={isInvalid}
        isValid={isValid}
        label={translate('onboarding.create_project.project_key')}
        required={true}>
        <input
          autoFocus={true}
          className={classNames('input-super-large', {
            'is-invalid': isInvalid,
            'is-valid': isValid
          })}
          id="project-key"
          maxLength={400}
          minLength={1}
          onChange={this.handleChange}
          type="text"
          value={this.state.value}
        />
      </ValidationInput>
    );
  }
}
