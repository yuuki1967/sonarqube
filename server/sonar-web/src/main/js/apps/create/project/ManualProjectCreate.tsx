/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as classNames from 'classnames';
import OrganizationInput from './OrganizationInput';
import { createProject } from '../../../api/components';
import DeferredSpinner from '../../../components/common/DeferredSpinner';
import { SubmitButton } from '../../../components/ui/buttons';
import ProjectKeyInput from '../components/ProjectKeyInput';
import VisibilitySelector from '../../../components/common/VisibilitySelector';
import UpgradeOrganizationBox from '../components/UpgradeOrganizationBox';
import HelpTooltip from '../../../components/controls/HelpTooltip';
import { translate } from '../../../helpers/l10n';
import { isSonarCloud } from '../../../helpers/system';
import './ManualProjectCreate.css';

interface Props {
  currentUser: T.LoggedInUser;
  fetchMyOrganizations?: () => Promise<void>;
  onProjectCreate: (projectKeys: string[]) => void;
  organization?: string;
  userOrganizations?: T.Organization[];
}

interface State {
  projectName?: string;
  projectNameChanged: boolean;
  projectKey?: string;
  selectedOrganization?: T.Organization;
  selectedVisibility?: T.Visibility;
  submitting: boolean;
}

type ValidState = State & Required<Pick<State, 'projectKey' | 'projectName'>>;

export default class ManualProjectCreate extends React.PureComponent<Props, State> {
  mounted = false;

  constructor(props: Props) {
    super(props);
    this.state = {
      projectNameChanged: false,
      selectedOrganization: this.getInitialSelectedOrganization(props),
      submitting: false
    };
  }

  componentDidMount() {
    this.mounted = true;
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  canChoosePrivate = (selectedOrganization: T.Organization | undefined) => {
    return Boolean(selectedOrganization && selectedOrganization.subscription === 'PAID');
  };

  canSubmit(state: State): state is ValidState {
    return Boolean(
      state.projectKey && state.projectName && (!isSonarCloud() || state.selectedOrganization)
    );
  }

  getInitialSelectedOrganization = (props: Props) => {
    if (props.organization) {
      return this.getOrganization(props.organization);
    } else if (props.userOrganizations && props.userOrganizations.length === 1) {
      return props.userOrganizations[0];
    } else {
      return undefined;
    }
  };

  getOrganization = (organizationKey: string) => {
    return (
      this.props.userOrganizations &&
      this.props.userOrganizations.find(({ key }: T.Organization) => key === organizationKey)
    );
  };

  handleFormSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const { state } = this;
    if (this.canSubmit(state)) {
      this.setState({ submitting: true });
      createProject({
        project: state.projectKey,
        name: state.projectName || state.projectKey,
        organization: state.selectedOrganization && state.selectedOrganization.key,
        visibility: this.state.selectedVisibility
      }).then(
        ({ project }) => this.props.onProjectCreate([project.key]),
        () => {
          if (this.mounted) {
            this.setState({ submitting: false });
          }
        }
      );
    }
  };

  handleOrganizationSelect = ({ key }: T.Organization) => {
    const selectedOrganization = this.getOrganization(key);
    let { selectedVisibility } = this.state;

    if (selectedVisibility === undefined) {
      selectedVisibility = this.canChoosePrivate(selectedOrganization) ? 'private' : 'public';
    }

    this.setState({
      selectedOrganization,
      selectedVisibility
    });
  };

  handleOrganizationUpgrade = () => {
    this.props.fetchMyOrganizations!().then(
      () => {
        this.setState(prevState => {
          if (prevState.selectedOrganization) {
            const selectedOrganization = this.getOrganization(prevState.selectedOrganization.key);
            return {
              selectedOrganization
            };
          }
          return null;
        });
      },
      () => {}
    );
  };

  handleProjectNameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const projectName = event.currentTarget.value.trim();
    this.setState({ projectName, projectNameChanged: true });
  };

  handleProjectKeyChange = (projectKey?: string) => {
    this.setState(state => ({
      projectKey,
      projectName: state.projectNameChanged ? state.projectName : projectKey || ''
    }));
  };

  handleVisibilityChange = (selectedVisibility: T.Visibility) => {
    this.setState({ selectedVisibility });
  };

  render() {
    const { selectedOrganization, submitting } = this.state;
    const canChoosePrivate = this.canChoosePrivate(selectedOrganization);

    return (
      <div className="create-project">
        <div className="flex-1 huge-spacer-right">
          <form className="manual-project-create" onSubmit={this.handleFormSubmit}>
            {isSonarCloud() &&
              this.props.userOrganizations && (
                <OrganizationInput
                  onChange={this.handleOrganizationSelect}
                  organization={selectedOrganization ? selectedOrganization.key : ''}
                  organizations={this.props.userOrganizations}
                />
              )}
            <ProjectKeyInput
              className="form-field"
              initialValue={this.state.projectKey}
              onChange={this.handleProjectKeyChange}
            />
            <div className="form-field">
              <label htmlFor="project-name">
                <span className="text-middle">
                  <strong>{translate('onboarding.create_project.display_name')}</strong>
                  <em className="mandatory">*</em>
                </span>
                <HelpTooltip
                  className="spacer-left"
                  overlay={translate('onboarding.create_project.display_name.help')}
                />
              </label>
              <div className="little-spacer-top spacer-bottom">
                <input
                  className={'input-super-large'}
                  id="project-name"
                  maxLength={255}
                  minLength={1}
                  onChange={this.handleProjectNameChange}
                  type="text"
                  value={this.state.projectName}
                />
              </div>
              <div className="note abs-width-400">
                {translate('onboarding.create_project.display_name.description')}
              </div>
            </div>
            {isSonarCloud() &&
              selectedOrganization && (
                <div
                  className={classNames('visibility-select-wrapper', {
                    open: Boolean(this.state.selectedOrganization)
                  })}>
                  <VisibilitySelector
                    canTurnToPrivate={canChoosePrivate}
                    onChange={this.handleVisibilityChange}
                    showDetails={true}
                    visibility={canChoosePrivate ? this.state.selectedVisibility : 'public'}
                  />
                </div>
              )}
            <SubmitButton disabled={!this.canSubmit(this.state) || submitting}>
              {translate('set_up')}
            </SubmitButton>
            <DeferredSpinner className="spacer-left" loading={submitting} />
          </form>
        </div>

        {isSonarCloud() &&
          selectedOrganization && (
            <div className="create-project-side-sticky">
              <UpgradeOrganizationBox
                className={classNames('animated', { open: !canChoosePrivate })}
                onOrganizationUpgrade={this.handleOrganizationUpgrade}
                organization={selectedOrganization}
              />
            </div>
          )}
      </div>
    );
  }
}
