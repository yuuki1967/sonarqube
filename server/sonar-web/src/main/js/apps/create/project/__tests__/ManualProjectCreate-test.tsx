/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import ManualProjectCreate from '../ManualProjectCreate';
import { change, submit, waitAndUpdate } from '../../../../helpers/testUtils';
import { createProject } from '../../../../api/components';

jest.mock('../../../../api/components', () => ({
  createProject: jest.fn().mockResolvedValue({ project: { key: 'bar', name: 'Bar' } })
}));

jest.mock('../../../../helpers/system', () => ({
  isSonarCloud: jest.fn().mockReturnValue(true)
}));

beforeEach(() => {
  (createProject as jest.Mock<any>).mockClear();
});

it('should render correctly', () => {
  expect(getWrapper()).toMatchSnapshot();
});

it('should correctly create a public project', async () => {
  const onProjectCreate = jest.fn();
  const wrapper = getWrapper({ onProjectCreate });
  wrapper.find('withRouter(OrganizationInput)').prop<Function>('onChange')({ key: 'foo' });

  change(wrapper.find('ProjectKeyInput'), 'bar');
  change(wrapper.find('input#project-name'), 'Bar');
  expect(wrapper.find('SubmitButton').prop('disabled')).toBe(false);

  submit(wrapper.find('form'));
  expect(createProject).toBeCalledWith({
    project: 'bar',
    name: 'Bar',
    organization: 'foo',
    visibility: 'public'
  });

  await waitAndUpdate(wrapper);
  expect(onProjectCreate).toBeCalledWith(['bar']);
});

it('should correctly create a private project', async () => {
  const onProjectCreate = jest.fn();
  const wrapper = getWrapper({ onProjectCreate });
  wrapper.find('withRouter(OrganizationInput)').prop<Function>('onChange')({ key: 'bar' });

  change(wrapper.find('ProjectKeyInput'), 'bar');
  change(wrapper.find('input#project-name'), 'Bar');

  submit(wrapper.find('form'));
  expect(createProject).toBeCalledWith({
    project: 'bar',
    name: 'Bar',
    organization: 'bar',
    visibility: 'private'
  });

  await waitAndUpdate(wrapper);
  expect(onProjectCreate).toBeCalledWith(['bar']);
});

function getWrapper(props = {}) {
  return shallow(
    <ManualProjectCreate
      currentUser={{ groups: [], isLoggedIn: true, login: 'foo', name: 'Foo', scmAccounts: [] }}
      fetchMyOrganizations={jest.fn()}
      onProjectCreate={jest.fn()}
      userOrganizations={[
        { key: 'foo', name: 'Foo' },
        { key: 'bar', name: 'Bar', subscription: 'PAID' }
      ]}
      {...props}
    />
  );
}
