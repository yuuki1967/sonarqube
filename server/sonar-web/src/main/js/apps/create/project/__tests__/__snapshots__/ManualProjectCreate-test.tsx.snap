// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="create-project"
>
  <div
    className="flex-1 huge-spacer-right"
  >
    <form
      className="manual-project-create"
      onSubmit={[Function]}
    >
      <withRouter(OrganizationInput)
        onChange={[Function]}
        organization=""
        organizations={
          Array [
            Object {
              "key": "foo",
              "name": "Foo",
            },
            Object {
              "key": "bar",
              "name": "Bar",
              "subscription": "PAID",
            },
          ]
        }
      />
      <ProjectKeyInput
        className="form-field"
        onChange={[Function]}
      />
      <div
        className="form-field"
      >
        <label
          htmlFor="project-name"
        >
          <span
            className="text-middle"
          >
            <strong>
              onboarding.create_project.display_name
            </strong>
            <em
              className="mandatory"
            >
              *
            </em>
          </span>
          <HelpTooltip
            className="spacer-left"
            overlay="onboarding.create_project.display_name.help"
          />
        </label>
        <div
          className="little-spacer-top spacer-bottom"
        >
          <input
            className="input-super-large"
            id="project-name"
            maxLength={255}
            minLength={1}
            onChange={[Function]}
            type="text"
          />
        </div>
        <div
          className="note abs-width-400"
        >
          onboarding.create_project.display_name.description
        </div>
      </div>
      <SubmitButton
        disabled={true}
      >
        set_up
      </SubmitButton>
      <DeferredSpinner
        className="spacer-left"
        loading={false}
        timeout={100}
      />
    </form>
  </div>
</div>
`;
